Tt 1626 Matches user based on cycle 3 attribute

SKeerthana opened the pull request:
I have chosen National insurance number as cycle 3 attribute for matching users.

Commit:
TT-1626 - Adds migration sql for national insurance number table
for cycle 3 attribute

Authors: @SKeerthana

Commit:
TT-1626 - Matches user based on cycle3 attributes when there is multiple
user matches in cycle1.

Just as an example of cycle3 attribute we have chosen National insurance
number.

Authors: @SKeerthana

Commit:
TT-1626 - Updating migration filenames after master rebase

Authors: @SKeerthana

Commit:
TT-1626 - Adds Database unit tests and integration tests

Authors: @SKeerthana

Commit:
TT-1626 - Adds match and no-match scenario JSON files for cycle3

Migration script inserts two users whose only difference is their
national insurance number

Authors: @SKeerthana

andy-paine commented on src/integration-test/java/uk/gov/ida/verifylocalmatchingserviceexample/MatchingServiceResourceTest.java:
```diff
@@ -89,6 +90,70 @@ public void shouldReturnMatchWhenUserWithCaseInsensitiveSurnameDateOfBirthPostco
+        String postcode = "some-postcode";
+
+        testDatabaseRule.ensureUserWithAddressExist(surname, firstName, dateOfBirth, postcode);
+        testDatabaseRule.ensureUserWithAddressExist(surname, firstName, dateOfBirth, postcode);
```
Duplicated line?

SKeerthana replied:
Thats because we want two users with same info except for national insurance number

andy-paine commented on src/main/java/uk/gov/ida/verifylocalmatchingserviceexample/service/Cycle1MatchingService.java:
```diff
@@ -54,6 +56,9 @@ public MatchStatusResponseDto matchUser(MatchingServiceRequestDto matchingServic
 
+        if(matchingUsers.size() > 1) {
+            return cycle3MatchingService.matchUser(matchingServiceRequest, matchingUsers);
+        }
```
Our decision about when to do Cycle 1 matching (when Cycle 0 fails) is in the `MatchingService`. I feel we should put the logic of when to do Cycle 3 matching in there too - it is a bit odd for the `Cycle1MatchingService` to deal with the `Cycle3MatchingService` directly.

andy-paine commented on src/main/java/uk/gov/ida/verifylocalmatchingserviceexample/service/Cycle3MatchingService.java:
```diff
@@ -0,0 +1,53 @@
+
+        if (matchingUserIds.size() == 1) {
+            verifiedPidDAO.save(matchingServiceRequest.getHashedPid(), matchingUserIds.get(0));
+            return MATCH;
```
Maybe we could commonise the concept of "matching a user" to do the insertion and return MATCH at the same time? It is 2 lines though so probably not needed

andy-paine replied:
I wouldn't bother with this upon further reflection - it would probably add complexity given how boilerplate-y Java is.

Commit:
TT-1626 - Refactoring to call Cycle3MatchingService from MatchingService
instead of Cycle1MatchingService

Also changes the migration to use MAX function to get last inserted serial
id since pg_get_serial_sequence doesn't work in H2

Auhtors: @SKeerthana

Commit:
TT-1626 - Updating the schema docs since we added nationalInsurance table
for cycle 3 matching

Authors: @SKeerthana
